src/desktop/space.rs

```diff
@@ -0,0 +1,419 @@
+                  // TODO: Only push surface damage
+                  //       But this would need to take subsurfaces into account and accumulate damage at least.
+                  //       Even better would be if damage would be ignored that is hidden by subsurfaces...
+              }*/
+            // Add damage for the new position (see TODO above for a better approach)
+            damage.push(geo);
+        }
+
+        // That is all completely new damage, which we need to store for subsequent renders
+        let new_damage = damage.clone();
+        // We now add old damage states, if we have an age value
+        if age > 0 && state.old_damage.len() >= age {
+            // We do not need older states anymore
+            state.old_damage.truncate(age);
+            damage.extend(state.old_damage.iter().flatten().copied());
+        } else {
+            // just damage everything, if we have no damage
+            damage = vec![output_geo];
+        }
+
+        // Optimize the damage for rendering
+        damage.retain(|rect| rect.overlaps(output_geo));
+        damage.retain(|rect| rect.size.h > 0 && rect.size.w > 0);
+        for rect in damage.clone().iter() {
+            // if this rect was already removed, because it was smaller as another one,
+            // there is no reason to evaluate this.
+            if damage.contains(rect) {
+                // remove every rectangle that is contained in this rectangle
+                damage.retain(|other| !rect.contains_rect(*other));
+            }
+        }
+
+        let output_transform: Transform = output.current_transform().into();
+        if let Err(err) = renderer.render(
+            output_transform
+                .transform_size(output_size)
+                .to_f64()
+                .to_physical(state.render_scale)
+                .to_i32_round(),
+            output_transform,
+            |renderer, frame| {
+                // First clear all damaged regions
+                for geo in &damage {
+                    slog::debug!(self.logger, "Clearing at {:?}", geo);
+                    frame.clear(
+                        clear_color,
+                        Some(geo.to_f64().to_physical(state.render_scale).to_i32_ceil()),
+                    )?;
+                }
+
+                // Then re-draw all window overlapping with a damage rect.
+                for window in self.windows.iter() {
+                    let wgeo = window_rect(window, &self.id);
+                    let mut loc = window_loc(window, &self.id);
+                    if damage.iter().any(|geo| wgeo.overlaps(*geo)) {
+                        loc -= output_geo.loc;
+                        slog::debug!(self.logger, "Rendering window at {:?}", wgeo);
+                        draw_window(renderer, frame, window, state.render_scale, loc, &self.logger)?;
+                        window_state(self.id, window).drawn = true;
+                    }
+                }
```

Comment:
I suppose the longer-term idea would be to only redraw the parts of the window that overlap with the damage spec, right?

Reply:
The code does that now. opaque_region is an optimization that I will likely not tackle inside this PR.